module/zcommon/zfs_fletcher.c

```diff
@@ -245,7 +214,107 @@ fletcher_4_incremental_byteswap(const void *buf, uint64_t size,
 	ZIO_SET_CHECKSUM(zcp, a, b, c, d);
 }
 
+static const struct fletcher_4_calls fletcher_4_generic_calls = {
+	.init = fletcher_4_generic_init,
+	.compute = fletcher_4_generic,
+	.compute_byteswap = fletcher_4_generic_byteswap,
+	.name = "generic"
+};
+
+static const struct fletcher_4_calls *chosen = &fletcher_4_generic_calls;
+
+void
+fletcher_4_native(const void *buf, uint64_t size, zio_cksum_t *zcp)
+{
+	chosen->init(zcp);
+	chosen->compute(buf, size, zcp);
+	if (chosen->fini != NULL)
+		chosen->fini(zcp);
```

Comment:
The difference is negligible. I compared them with a patch as follows:
```
@@ -211,9 +263,31 @@ main(int argc, char **argv)
        fletcher_4_native(buf, size, &zc_ref);
        fletcher_4_byteswap(buf, size, &zc_ref_byteswap);
 
-       fletcher_4_avx2_init(&zc);
-       fletcher_4_avx2(buf, size, &zc);
-       fletcher_4_avx2_fini(&zc);
+       struct timespec s1, e1, s2, e2;
+
+#include <time.h>
+       i = 0;
+       clock_gettime(CLOCK_PROCESS_CPUTIME_ID, &s1);
+       while (i++ < 1000) {
+               fletcher_4_avx2_init(&zc);
+               fletcher_4_avx2(buf, size, &zc);
+               fletcher_4_avx2_fini(&zc);
+       }
+       clock_gettime(CLOCK_PROCESS_CPUTIME_ID, &e1);
+
+       i = 0;
+       clock_gettime(CLOCK_PROCESS_CPUTIME_ID, &s2);
+       while (i++ < 1000) {
+               fletcher_4_avx2_all_in_one(buf, size, &zc);
+       }
+       clock_gettime(CLOCK_PROCESS_CPUTIME_ID, &e2);
+
+       printf("buffer size: %zu bytes, loop 1000 times, %dms vs. %dms\n", size,
+               ((e1.tv_sec - s1.tv_sec) * 1000000000 + (int)(e1.tv_nsec - s1.tv_nsec)) / 1000000,
+               ((e2.tv_sec - s2.tv_sec) * 1000000000 + (int)(e2.tv_nsec - s2.tv_nsec))/1000000);
+
+       exit(0);
```
And the results are:

```
[root@centos7 zcommon]# ./zfs_fletcher_intel 
buffer size: 5614080 bytes, loop 1000 times, 342ms vs. 318ms
[root@centos7 zcommon]# ./zfs_fletcher_intel 
buffer size: 635904 bytes, loop 1000 times, 37ms vs. 36ms
[root@centos7 zcommon]# ./zfs_fletcher_intel 
buffer size: 11104256 bytes, loop 1000 times, 683ms vs. 681ms
```